pkg/skaffold/deploy/kpt/kpt.go

```diff
@@ -548,9 +478,7 @@ func getResources(root string) ([]string, error) {
 
 // getKptFnRunArgs returns a list of arguments that the user specified for the `kpt fn run` command.
 func (k *Deployer) getKptFnRunArgs() ([]string, error) {
-	// --dry-run sets the pipeline's output to STDOUT, otherwise output is set to sinkDir.
-	// For now, k.Dir will be treated as sinkDir (and sourceDir).
-	flags := []string{"--dry-run"}
```

Comment:
Sorry, I forgot to mention this in the description. This is another workaround that is no longer relevant, kpt will indeed output the to stdout without the `--dry-run` flag. I have verified this using the kpt version that I bumped the requirement to.

Reply:
> > Hi Mattias, Thank you so much for improving kpt deployer! I didn't realize the workaround is no longer needed (and actually add bugs).
> > I have several questions want to confirm:
> > 
> > 1. kpt and kustomize are decided to not be unified for now shrug . Thus the directories of kustomize and kpt are better not be overlapped when used together. Technically they can, this requires users to handle the potential issues themselves and better use kustomize as a kpt fn (and we will support that in the skaffold v2). So it seems the "removing comments" and "the kpt relative file breakage" are less of a concern.
> 
> I don't understand what you mean with "not unified" and that "directories should not be overlapped". Do you mean that one shouldn't really use both? That would contradict the next item a bit where you explain why you think the existing order is the right one. Could you also elaborate on why you think "removing comments" and "the kpt relative file breakage" are less of a concern? Especially the former is a blocker for us.
> 
Unified --> See [comment](https://github.com/GoogleContainerTools/skaffold/pull/5886/files#diff-ff7acade03900f504cf09e15ca05d1b444770acab8cd6ec7fd62496182b632a9R275) basically once they are unified, kpt can understand and work with kustomize config by default.  I just want to make sure the previous comment is not misleading.
 
The new kpt v1.0 is going to update the hydrated config in-place (the config is updated in its original file)  while kustomize outputs the config to stdout. From that perspective, the `kustomize build --DIR` and `kpt fn DIR` may mess up if the DIR are the same. So a clean approach would be make sure kpt dir and kustomize dir do not overlap. A useful but may not clean approach would be re-creating and applying the setters after `kustomize build`, which allows using kustomize config and kpt-setters in the same dir.

> Running kustomize as a kpt fn sounds very interesting though. Can you provide any more information on this? I assume that is not going to be a Skaffold specific feature but rather something that Kustomize will support, or?
> 
> > 1. `kustomize build` is considered as the generator, and kpt fn provides the functions to validate and transform these resources. So the order is 1) generate, 2) validate 3) mutate. From this perspective, ignoring whether doing kustomize first may break kpt features or vice versa, running `kustomize build` first seems to be the right approach. And if anything's broken during the process, we shall fix the breakage.
> 

Here's [the issue](https://github.com/GoogleContainerTools/kpt/issues/1447). But we've agreed to have a skaffold specific kustomize kpt fn rather than a generic one (which needs to handle all possible use cases and we don't have many user requests yet).  

> Kpt functions that are validators and should run last is a fair point. For me the ability to use functions that require comments and the ability for the functions to modify Kustomizations is more important, but I realize that this will not apply to everybody. If you believe that changing the order is the wrong thing to do then I'll drop this PR.
> 


> > 1. Does this PR fix the bug that "using --fn-path will cause function configs to be used only from the specified path whereas the workaround added those to whatever was in kpt.dir"? Maybe I miss something but it seems that this issue still exists. Do you mind updating the PR to filter out the kptFn from `.kpt.dir` if `--fnPath` is given? somewhere after the [kpt source](https://github.com/GoogleContainerTools/skaffold/pull/5886/files#diff-ff7acade03900f504cf09e15ca05d1b444770acab8cd6ec7fd62496182b632a9R260).
> 
> No, but I don't think this is a bug, `kpt fn run --help` explicitly says this about `--fn-path`: `read functions from these directories instead of the configuration directory.` In my opinion it is unexpected that Skaffold tries to change intended behavior of Kpt. The way I read the existing code and the comments was that the function configs in `kpt.fn.fnPath` were added to `kpt.dir` as a workaround for `--fn-path` blowing up and I assumed that the additive behavior was an unintended side effect.
> 

The expected behavior of skaffold kpt deployer should be: If `.kpt.fn.fnPath` is not given, skaffold searches and applies the kpt functions from `.kpt.dir` (declarative kpt functions); If `.kpt.fn.fnPath` is given, skaffold should only apply these functions and intentionally filter out the kpt functions defined in `.kpt.dir`. The existing code does not filter out the fn from `.kpt.dir`   

> > 1. I feel sorry that the kpt deployer breaks. The kpt v0.X.X are not guaranteed to be backward compatible (but kpt v1.0 will do much better on this). Thus, instead of fixing kpt deployer in skaffold everytime it breaks (I've done similar fixes a couple of times), do you think it may actually make users life easier if we require a specific kpt version? e.g. here we ask users to install kpt 0.34.0  and do not upgrade to newer kpt 0.X.XX versions until the skaffold kpt v1.0 is ready (I'm currently implementing the new kpt hydration/deployment based on kpt v1.0).
> 
> I'm not sure if that would make things better, it's also pretty inconvenient to not be able to upgrade to the latest version for other use cases.
yeah, and this is the common problem for most non-GA (v1.0) CLIs.